File: protocol/impls/py/prng.py
"""SplitMix64 PRNG (SPEC-conformance.md §1) — self-implemented, stdlib only.

All arithmetic wraps mod 2^64 (Python int is unbounded; mask every step).
"""

MASK64 = (1 << 64) - 1
GAMMA = 0x9E3779B97F4A7C15
M1 = 0xBF58476D1CE4E5B9
M2 = 0x94D049BB133111EB


class SplitMix64:
    def __init__(self, seed):
        # "the seed IS the state (no warm-up)"
        self.state = seed & MASK64

    def next(self):
        self.state = (self.state + GAMMA) & MASK64
        z = self.state
        z = ((z ^ (z >> 30)) * M1) & MASK64
        z = ((z ^ (z >> 27)) * M2) & MASK64
        return z ^ (z >> 31)

    def bounded(self, n):
        # "n>0; n==0 -> 0. PINNED as a plain modulo."
        if n == 0:
            return 0
        return self.next() % n


def _selftest():
    r = SplitMix64(0)
    first = r.next()
    assert first == 0xE220A8397B1DCDAF, hex(first)


if __name__ == "__main__":
    _selftest()
    print("prng selftest ok")

File: protocol/impls/rust/src/prng.rs
//! SplitMix64 PRNG (pinned, §1 of SPEC-conformance.md). All wrapping u64.
//! Conformance: next() from seed=0 returns 0xE220A8397B1DCDAF first.

pub struct SplitMix64 {
    state: u64,
}

impl SplitMix64 {
    pub fn new(seed: u64) -> Self {
        // "the seed IS the state (no warm-up)"
        SplitMix64 { state: seed }
    }

    pub fn next(&mut self) -> u64 {
        self.state = self.state.wrapping_add(0x9E3779B97F4A7C15);
        let mut z = self.state;
        z = (z ^ (z >> 30)).wrapping_mul(0xBF58476D1CE4E5B9);
        z = (z ^ (z >> 27)).wrapping_mul(0x94D049BB133111EB);
        z ^ (z >> 31)
    }

    /// bounded(n) := next() mod n ; n==0 → 0. PINNED as a plain modulo.
    pub fn bounded(&mut self, n: u64) -> u64 {
        if n == 0 {
            0
        } else {
            self.next() % n
        }
    }
}
